### ISSUE TITLE
Failure to evaluate network with plotting on with large training sets

### ISSUE THREAD

**Issue by backyardbiomech** [DeepLabCut/DeepLabCut#910]
Related to one of the sub-issues in #878, and to #835 , but I think worthy of elevating to its own issue since both of those other issues covered multiple issues, and this one was never solved (just worked around). I've tried to monitor a few things (details below).

**Describe the bug**

During evaluating network with plotting, I get a wx._core.wxAssertionError (the same as mentioned [here](https://github.com/DeepLabCut/DeepLabCut/issues/878#issuecomment-683074629) and at the start of #835 . Some but not all plots are saved, and no other evaluation outputs are returned.

**Desktop (please complete the following information about your system):**
 - OS: Windows 10, 64 GB, 24 GB GPU
 - DeepLabCut Version: 2.2b8 - clean install of conda env 15 Sept.
 - Browser, if applicable [e.g. chrome, safari]

**To Reproduce**
Same with GUI or from command line:
1. Create training dataset with large number of input images, and crop and label True.
2. Train Network
3. Evaluate Network with plotting on

<details><summary>TRACEBACK</summary>
<p>

```python

1219it [05:34,  3.64it/s]
Traceback (most recent call last):
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\gui\evaluate_network.py", line 340, in evaluate_network
    comparisonbodypart
s=self.bodyparts,
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\pose_estimation_tensorflow\evaluate.py", line 583, in evaluate_network
    c_engine=c_engine,
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\pose_estimation_tensorflow\evaluate_multianimal.py", line 343, in evaluate_multianimal_full
    cfg["pcutoff"],
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\utils\visualization.py", line 109, in make_multianimal_labeled_image
    fig, ax = prepare_figure_axes(w, h)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\utils\visualization.py", line 185, in prepare_figure_axes
    frameon=False, figsize=(width * scale / dpi, height * scale / dpi), dpi=dpi
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\pyplot.py", line 545, in figure
    **kwargs)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1969, in new_figure_manager
    return super().new_figure_manager(num, *args, **kwargs)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backend_bases.py", line 3261, in new_figure_manager
    return cls.new_figure_manager_given_figure(num, fig)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1973, in new_figure_manager_given_figure
    frame = cls._frame_class(num, figure)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1103, in __init__
    self.toolbar = self._get_toolbar(self.statusbar)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1131, in _get_toolbar
    toolbar = NavigationToolbar2Wx(self.canvas)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1377, in __init__
    NavigationToolbar2.__init__(self, canvas)
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backend_bases.py", line 2636, in __init__
    self._init_toolbar()
  File "C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\matplotlib\backends\backend_wx.py", line 1413, in _init_toolbar
    self.Realize()
wx._core.wxAssertionError: C++ assertion "Assert failure" failed at ..\..\src\msw\toolbar.cpp(938) in wxToolBar::Realize(): Could not add bitmap to toolbar  


```

Also, possible relevant proceeding FutureWarnings and RuntimeWarnings. There are MANY of these, I'm assuming related to cropped training images when all points have been cropped out. But maybe not trying to save these would prevent memory issues?
```python
1216it [04:52,  3.84it/s]C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\pose_estimation_tensorflow\evaluate_multianimal.py:266: FutureWarning: Pass-through of possibly RGB images in gray2rgb is deprecated. In version 0.19, input arrays will always be considered grayscale, even if the last dimension has length 3 or 4. To prevent this warning and ensure compatibility with future versions, detect RGB images outside of this function.
  frame = img_as_ubyte(skimage.color.gray2rgb(image))
C:\Users\jacksonbe3\Miniconda3\envs\dlc\lib\site-packages\deeplabcut\pose_estimation_tensorflow\evaluate_multianimal.py:280: RuntimeWarning: All-NaN slice encountered
  distnorm[imageindex] = np.nanmax(lengths)
```
</p></details>

**Expected behavior**
Evaluated images from all images in the training set, and other evaluation outputs

**Screenshots**
If applicable, add screenshots to help explain your problem. (but please post the code output above!)

**Additional context**
Many labeled images are produced (ca. 1215), but with a large number of labelled images, and "Crop and Label" selected when creating the the dataset, this isn't all of them. Therefore, it acts like a memory leak issue, also suggested by and with a possible strategy for a fix [here](https://stackoverflow.com/a/61920189).

Task Manager shows a gradual ongoing drop in available RAM, with this process eventually taking up about 2.5 GB as the evaluate process continues, so it doesn't seem to be releasing memory after creating each figure. However, when it fails, I still have about 45 GB free, so if it is a memory problem it's a problem with pre-allocation and not total available memory. In fact, that memory isn't released even with the error and the process stopping; it is only released if I start another evaluation process (or close the GUI).

There isn't an error with smaller training sets with crop and label. or the same project without crop and label, or this data set but without plotting, so it doesn't seem to be something fundamental to the version of wxpython, but indeed something about memory handling.

Importantly, with plotting off, evaluation completes and returns everything (other than the images).

**Comment by CalvinLloyd117** [comment]
I too am experiencing a very similar issue to the one described above. I have ~20 labelled frames from each of the 10 different videos for training. After using the crop and label, there was around 2000 cropped frames (400x400 if i recall correctly). Eventually, the process crashes with the following error:

wx._core.wxAssertionError: C++ assertion "m_hdc" failed at ..\..\src\msw\textmeasure.cpp(64) in wxTextMeasure::BeginMeasuring(): Must not be used with non-native wxDCs 

This is occurring in Windows 10, and deeplabcut==2.2b7.

As described above, with *Want to plot predictions (as in standard DLC projects)* set to no, the process completes as normal, but without the images.

**Comment by MMathisLab** [comment]
can you both confirm you are using <4.1.0 wxPython?

`import wx`
`wx.__version__`

**Comment by backyardbiomech** [comment]
```
import wx
wx.__version__
'4.0.4'
```

and for other packages if it helps...
<details>
<summary>output of `conda list`</summary>

```python
# Name                    Version                   Build  Channel
_tflow_select             2.1.0                       gpu
absl-py                   0.9.0                    py37_0
argon2-cffi               20.1.0           py37he774522_1
astor                     0.8.1                    py37_0
attrs                     20.1.0                     py_0
backcall                  0.2.0                      py_0
bayesian-optimization     1.2.0                    pypi_0    pypi
blas                      1.0                         mkl
bleach                    3.1.5                      py_0
ca-certificates           2020.7.22                     0
certifi                   2020.6.20                py37_0
cffi                      1.14.2           py37h7a1dbc1_0
chardet                   3.0.4                    pypi_0    pypi
click                     7.1.2                    pypi_0    pypi
colorama                  0.4.3                      py_0
cudatoolkit               10.0.130                      0
cudnn                     7.6.5                cuda10.0_0
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.21                  pypi_0    pypi
decorator                 4.4.2                      py_0
deeplabcut                2.2b8                    pypi_0    pypi
defusedxml                0.6.0                      py_0
easydict                  1.9                      pypi_0    pypi
entrypoints               0.3                      py37_0
filterpy                  1.4.5                    pypi_0    pypi
gast                      0.4.0                      py_0
geos                      3.8.0                h33f27b4_0
grpcio                    1.31.0           py37he7da953_0
h5py                      2.10.0           py37h5e291fa_0
hdf5                      1.10.4               h7ebc959_0
icc_rt                    2019.0.0             h0cc432a_1
icu                       58.2                 ha925a31_3
idna                      2.10                     pypi_0    pypi
imageio                   2.9.0                    pypi_0    pypi
imageio-ffmpeg            0.4.2                    pypi_0    pypi
imgaug                    0.4.0                    pypi_0    pypi
importlib-metadata        1.7.0                    py37_0
importlib_metadata        1.7.0                         0
intel-openmp              2019.0                   pypi_0    pypi
ipykernel                 5.3.4            py37h5ca1d4c_0
ipython                   7.18.1           py37h5ca1d4c_0
ipython_genutils          0.2.0                    py37_0
ipywidgets                7.5.1                      py_0
jedi                      0.17.2                   py37_0
jinja2                    2.11.2                     py_0
joblib                    0.16.0                   pypi_0    pypi
jpeg                      9b                   hb83a4c4_2
jsonschema                3.2.0                    py37_1
jupyter                   1.0.0                    py37_7
jupyter_client            6.1.6                      py_0
jupyter_console           6.2.0                      py_0
jupyter_core              4.6.3                    py37_0
keras-applications        1.0.8                      py_1
keras-preprocessing       1.1.0                      py_1
kiwisolver                1.2.0                    pypi_0    pypi
libpng                    1.6.37               h2a8f88b_0
libprotobuf               3.13.0               h200bbdf_0
libsodium                 1.0.18               h62dcd97_0
llvmlite                  0.34.0                   pypi_0    pypi
m2w64-gcc-libgfortran     5.3.0                         6
m2w64-gcc-libs            5.3.0                         7
m2w64-gcc-libs-core       5.3.0                         7
m2w64-gmp                 6.1.0                         2
m2w64-libwinpthread-git   5.0.0.4634.697f757               2
markdown                  3.2.2                    py37_0
markupsafe                1.1.1            py37hfa6e2cd_1
matplotlib                3.1.3                    pypi_0    pypi
mistune                   0.8.4           py37hfa6e2cd_1001
mkl                       2020.2                      256
mkl-service               2.3.0            py37hb782905_0
mkl_fft                   1.1.0            py37h45dec08_0
mkl_random                1.1.1            py37h47e9c7a_0
mock                      4.0.2                      py_0
moviepy                   1.0.1                    pypi_0    pypi
msgpack                   1.0.0                    pypi_0    pypi
msgpack-numpy             0.4.6.1                  pypi_0    pypi
msys2-conda-epoch         20160418                      1
nb_conda                  2.2.1                    py37_0
nb_conda_kernels          2.2.4                    py37_0
nbconvert                 5.6.1                    py37_1
nbformat                  5.0.7                      py_0
networkx                  2.5                      pypi_0    pypi
notebook                  6.1.1                    py37_0
numba                     0.51.1                   pypi_0    pypi
numexpr                   2.7.1                    pypi_0    pypi
numpy                     1.16.4                   pypi_0    pypi
numpy-base                1.19.1           py37ha3acd2a_0
opencv-python             4.4.0.42                 pypi_0    pypi
opencv-python-headless    4.4.0.42                 pypi_0    pypi
openssl                   1.1.1g               he774522_1
packaging                 20.4                       py_0
pandas                    1.1.2                    pypi_0    pypi
pandoc                    2.10.1                        0
pandocfilters             1.4.2                    py37_1
parso                     0.7.0                      py_0
patsy                     0.5.1                    pypi_0    pypi
pickleshare               0.7.5                 py37_1001
pillow                    7.2.0                    pypi_0    pypi
pip                       20.2.2                   py37_0
proglog                   0.1.9                    pypi_0    pypi
prometheus_client         0.8.0                      py_0
prompt-toolkit            3.0.7                      py_0
prompt_toolkit            3.0.7                         0
protobuf                  3.13.0           py37h6538335_0
psutil                    5.7.2                    pypi_0    pypi
pycparser                 2.20                       py_2
pygments                  2.7.0                      py_0
pyparsing                 2.4.7                      py_0
pyqt                      5.9.2            py37h6538335_2
pyreadline                2.1                      py37_1
pyrsistent                0.16.0           py37he774522_0
python                    3.7.9                h60c2a47_0
python-dateutil           2.8.1                      py_0
pytz                      2020.1                   pypi_0    pypi
pywavelets                1.1.1                    pypi_0    pypi
pywin32                   227              py37he774522_1
pywinpty                  0.5.7                    py37_0
pyyaml                    5.3.1                    pypi_0    pypi
pyzmq                     19.0.1           py37ha925a31_1
qt                        5.9.7            vc14h73c81de_0
qtconsole                 4.7.6                      py_0
qtpy                      1.9.0                      py_0
requests                  2.24.0                   pypi_0    pypi
ruamel-yaml               0.16.12                  pypi_0    pypi
ruamel-yaml-clib          0.2.2                    pypi_0    pypi
scikit-image              0.17.2                   pypi_0    pypi
scikit-learn              0.23.2                   pypi_0    pypi
scipy                     1.5.2            py37h9439919_0
send2trash                1.5.0                    py37_0
setuptools                49.6.0                   py37_0
shapely                   1.7.1            py37h210f175_0
sip                       4.19.8           py37h6538335_0
six                       1.15.0                     py_0
sqlite                    3.33.0               h2a8f88b_0
statsmodels               0.12.0                   pypi_0    pypi
tables                    3.6.1                    pypi_0    pypi
tabulate                  0.8.7                    pypi_0    pypi
tensorboard               1.13.1           py37h33f27b4_0
tensorflow                1.13.1          gpu_py37h83e5d6a_0
tensorflow-base           1.13.1          gpu_py37h871c8ca_0
tensorflow-estimator      1.13.0                     py_0
tensorflow-gpu            1.13.1               h0d30ee6_0
tensorpack                0.9.8                    pypi_0    pypi
termcolor                 1.1.0                    py37_1
terminado                 0.8.3                    py37_0
testpath                  0.4.4                      py_0
threadpoolctl             2.1.0                    pypi_0    pypi
tifffile                  2020.9.3                 pypi_0    pypi
tornado                   6.0.4            py37he774522_1
tqdm                      4.49.0                   pypi_0    pypi
traitlets                 4.3.3                    py37_0
urllib3                   1.25.10                  pypi_0    pypi
vc                        14.1                 h0510ff6_4
vs2015_runtime            14.16.27012          hf0eaf9b_3
wcwidth                   0.2.5                      py_0
webencodings              0.5.1                    py37_1
werkzeug                  1.0.1                      py_0
wheel                     0.35.1                     py_0
widgetsnbextension        3.5.1                    py37_0
wincertstore              0.2                      py37_0
winpty                    0.4.3                         4
wxpython                  4.0.4            py37ha925a31_0
zeromq                    4.3.2                ha925a31_3
zipp                      3.1.0                      py_0
zlib                      1.2.11               h62dcd97_4
```
</details>

**Comment by CalvinLloyd117** [comment]
```
In [1]: import wx
In [2]: wx.__version__
Out[2]: '4.0.7'
```

**Comment by backyardbiomech** [comment]
On my latest attempt to plot everything during evaluation, windows task manager is now showing the memory leak issue whereas it didn't show it before. Note that this is NOT a cropped training data set (higher resolution images though).

![image](https://user-images.githubusercontent.com/9662673/94351045-909c9100-0022-11eb-9c31-7d15997d4834.png).

Once it gets to less than 100 MB of available RAM, 1-3 GB get cleared and then refill. It looks like the clearing is only happening by adding committed (virtual) memory, and the plotting processes are not releasing any memory. If you compare this screenshot with the previous you can see a large increase in committed memory.

![image](https://user-images.githubusercontent.com/9662673/94351124-93e44c80-0023-11eb-83c6-fd2ec9e490b4.png)

My guess is the crash eventually occurs when it runs out of space for virtual memory on my SSD.

**Comment by jeylau** [comment]
@backyardbiomech, this will get fixed soon 😊

**Comment by MMathisLab** [comment]
Will be fixed in #1208! Thanks for the patience on this one 👯
